Client/Source/Client.cpp

```diff
@@ -0,0 +1,361 @@
+
+    connect(_socketForLayer.get(), &QTcpSocket::readyRead, this, &Client::slotReadFromLayer);
+
+    connect(_socketForLayer.get(), &QTcpSocket::disconnected, this, &Client::slotServerDisconnected,
+            Qt::QueuedConnection);
+
+    connect(this, &Client::signalToSend, this, &Client::slotSendDataToLayer);
+}
+
+Client::Client(const int serverReceivingPort, const int serverSendingPort,
+               const std::string_view serverIP, const WorkMode workMode, QObject* parent)
+    : QObject(parent),
+      _layerPort(0),
+      _socketForLayer(nullptr),
+      _serverReceivingPort(serverReceivingPort),
+      _receivingSocket(std::make_unique<QTcpSocket>(this)),
+      _serverSendingPort(serverSendingPort),
+      _sendingSocket(std::make_unique<QTcpSocket>(this)),
+      _serverIP(serverIP),
+      _start(std::chrono::steady_clock::now()),
+      _workMode(workMode),
+      _logger(CONFIG.get<CAST(Param::DEFAULT_IN_FILE_NAME)>(),
+              CONFIG.get<CAST(Param::DEFAULT_OUT_FILE_NAME)>())
+{
+    _printer.writeLine(std::cout, "Server Receiving Port:", serverReceivingPort,
+                       "Server Sending Port:", serverSendingPort, "Server IP:", serverIP);
+
+    connect(_receivingSocket.get(), &QTcpSocket::readyRead, this, &Client::slotReadFromServer);
+    connect(_receivingSocket.get(), &QTcpSocket::disconnected, this,
+            &Client::slotServerDisconnected, Qt::QueuedConnection);
+
+    ///connect(_sendingSocket.get(), &QTcpSocket::readyRead, this, &Client::slotClientRead);
+    connect(_sendingSocket.get(), &QTcpSocket::disconnected, this, &Client::slotServerDisconnected,
+            Qt::QueuedConnection);
+
+    connect(this, &Client::signalToSend, this, &Client::slotSendDataToServer);
+}
+
+void Client::slotReadFromLayer()
+{
+    if (_socketForLayer->bytesAvailable() > 0)
+    {
+        const QByteArray array = _socketForLayer->readAll();
+        const std::string receivedData = array.toStdString();
+
+        _duration = std::chrono::steady_clock::now() - _start;
+        ///_start = std::chrono::steady_clock::now();
+
+        _printer.writeLine(std::cout, _socketForLayer->localPort(), '-', receivedData);
+        _logger.writeLine(_socketForLayer->localPort(), '-', receivedData);
+
+        _printer.writeLine(std::cout, "Duration:", _duration.count(), "seconds");
+        _logger.writeLine("Duration:", _duration.count(), "seconds");
+    }
+}
+
+void Client::slotReadFromServer()
+{
+    if (_receivingSocket->bytesAvailable() > 0)
+    {
+        const QByteArray array = _receivingSocket->readAll();
+        const std::string receivedData = array.toStdString();
+
+        _duration = std::chrono::steady_clock::now() - _start;
+        ///_start = std::chrono::steady_clock::now();
+
+        _printer.writeLine(std::cout, _receivingSocket->localPort(), '-', receivedData);
+        _logger.writeLine(_receivingSocket->localPort(), '-', receivedData);
```

Comment:
**(note, out of scope): ** May be it will be a good idea to combine the methods, because "logger" - it is same printer, but destination is file.

Reply:
Firstly, not every logging message we need to show to user. Secondly, `Printer` and `Logger` are not the same classes. The first one is singleton, but the second one is not. And every module should have own `Logger` instance. But work with terminal is common for all components in one module.